rule/doc.go

```diff
@@ -0,0 +1,153 @@
+// operation will enforce the contract as we go.  The return type is
+// also Expr, because the rule's abstract syntax tree will be a
+// homogeneous tree of Exprs.
+//
+// 🛈 Every one of these public constructors will end up looking the
+// same, with only the private constructor they use varying.  This is
+// probably a candidate for code generation in the future.
+//
+//    func Plus(vN ...Expr) Expr {
+//            e := newExprPlus()
+//            e.consumeOperands(vN...)
+//            return e
+//    }
+//
+// To make the operator do anything at all, we have to define its
+// evaluator. Our "plus" operator has to add things together!  Our
+// Contract says we'll add NUMBER types, which resolve to either
+// Float64Value or Int64Value in Regula.  We specifically don't have
+// to worry about mixed addition of these types because our Contract
+// says we want MANY of type NUMBER, and we can rely on the parser to
+// promote INTEGER to FLOAT if we pass a mixed list of NUMBER.  To
+// simplify the code for the two concrete types, I've split them out
+// of the public Eval method.
+//
+//    func (n *exprPlus) int64Sum(params) (*Value, error) {
+//            var sum int64
+//
+//            for _, op := range n.operands {
+//                    val, err := op.Eval(params)
+//                    if err != nil {
+//                            return nil, err
+//                    }
+//                    ival, err != strconv.Atoi(val)
+//                    if err != nil {
+//                            return nil, err
+//                    }
+//                    sum += ival
+//            }
+//            return Int64Value(sum), nil
+//    }
+//
+//    func (n *exprPlus) float64Sum(params) (*Value, error) {
+//            var sum float64
+//
+//            for _, op := range n.operands {
+//                    val, err := op.Eval(params)
+//                    if err != nil {
+//                            return nil, err
+//                    }
+//                    fval, err != strconv.ParseFloat(val, 64)
+//                    if err != nil {
+//                            return nil, err
+//                    }
+//                    sum += fval
+//            }
+//            return Int64Value(sum), nil
+//    }
+//
+//    func (n *exprPlus) Eval(params Params) (*Value, error) {
+//            typ := n.operands[0].Type
+//            switch typ {
+//            case "int64":
+//                    return n.int64Sum(params)
+//            case "float64":
+//                    return n.float64Sum(params)
+//            }
+//            return nil, fmt.Errorf("Invalid NUMBER type %q", typ)
+//    }
+//
+//
+// Finally, we need a way for people using language other than Go to
+// construct our Operator.  The Operator interface defines most of
+// what we need, bus we'll need to modify the GetOperator function in
+// operator.go to reflect our new operators existence.
```

Comment:
What is the relation between those two sentences: 
> Finally, we need a way for people using language other than Go to construct our Operator.

and

> The Operator interface defines most of what we need, bus we'll need to modify the GetOperator function in operator.go to reflect our new operators existence.

Reply:
Not sure to understand the meaning here.

Reply:
It's reflecting that we don't have a mechanism, inherent to Go, to refer to a a Go function by a string containing its name, so we have to maintain a static mapping from an OpCode (name as a string) to an actual constructor function.  Currently it's a switch statement, in the future it'll probably need to be a map (to support manipulating it at run time).